src/Databases/DatabasesCommon.cpp

```diff
@@ -31,8 +35,65 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
     extern const int CANNOT_GET_CREATE_TABLE_QUERY;
 }
+namespace
+{
+void validateCreateQuery(const ASTCreateQuery & query, ContextPtr context)
+{
+    /// First validate that the query can be parsed
+    const auto serialized_query = serializeAST(query);
+    ParserCreateQuery parser;
+    ASTPtr new_query_raw = parseQuery(
+        parser,
+        serialized_query.data(),
+        serialized_query.data() + serialized_query.size(),
+        "after altering table ",
+        0,
+        context->getSettingsRef().max_parser_depth,
+        context->getSettingsRef().max_parser_backtracks);
+    const auto & new_query = new_query_raw->as<const ASTCreateQuery &>();
+    /// If there are no columns, then there is nothing much we can do
+    if (!new_query.columns_list || !new_query.columns_list->columns)
+        return;
+
+    const auto & columns = *new_query.columns_list;
+    /// Do some basic sanity checks. Let's enforce strict rules, just like on CREATE, because otherwise the default expressions might not be checked
+    const auto columns_desc
+        = InterpreterCreateQuery::getColumnsDescription(*columns.columns, context, LoadingStrictnessLevel::CREATE, false);
```

Comment:
This is something I am not happy about, but don't see any other easy way. One solution might be to extract the logic from `getColumnsDescription` that creates `default_expr_list`, then `validateColumnsDefaultsAndGetSampleBlock` could be called outside of `getColumnsDescription`. What do you think, does it worth to extract the logic, or just use `CREATE` here?

Reply:
This is fine as is

Reply:
I ended up separating it, because:
1. `getColumnsDescription` with `LoadingStrictnessLevel::CREATE` does a lot of checks that might fail without the correct context (e.g. session settings allow some suspicious/experimental thing), so I had to lower the level to `SECONDARY_CREATE`. But that doesn't check the default expressions. I tried to only validate, but not evaluate the default expressions, but that breaks `ATTACH` queries. So I ended up separating the logic that is necessary to validate default expressions.